src/adapters/source-adapters/BoundedUnionSourceAdapter.sol

```diff
@@ -91,21 +104,46 @@ abstract contract BoundedUnionSourceAdapter is
         override(ChainlinkSourceAdapter, ChronicleMedianSourceAdapter, PythSourceAdapter)
         returns (int256, uint256, uint256)
     {
-        // Chainlink has price history, so use tryLatestDataAt to pull the most recent price that satisfies the timestamp constraint.
-        (int256 clAnswer, uint256 clTimestamp,) = ChainlinkSourceAdapter.tryLatestDataAt(timestamp, maxTraversal);
+        // In the happy path there have been no source updates since requested time, so we can return the latest data.
+        AllSourceData memory data = _getAllLatestSourceData();
+        (int256 boundedAnswer, uint256 boundedTimestamp) = _selectBoundedPrice(
+            data.clAnswer, data.clTimestamp, data.crAnswer, data.crTimestamp, data.pyAnswer, data.pyTimestamp
+        );
+        if (boundedTimestamp <= timestamp) return (boundedAnswer, boundedTimestamp, 1);
 
-        // For Chronicle and Pyth, just pull the most recent prices and drop them if they don't satisfy the constraint.
-        (int256 crAnswer, uint256 crTimestamp) = ChronicleMedianSourceAdapter.getLatestSourceData();
-        (int256 pyAnswer, uint256 pyTimestamp) = PythSourceAdapter.getLatestSourceData();
+        // Chainlink has price history, so use tryLatestDataAt to pull the most recent price that satisfies the timestamp constraint.
+        (data.clAnswer, data.clTimestamp,) = ChainlinkSourceAdapter.tryLatestDataAt(timestamp, maxTraversal);
 
-        // To "drop" Chronicle and Pyth, we set their timestamps to 0 (as old as possible) if they are too recent.
+        // "Drop" Chronicle and/or Pyth by setting their timestamps to 0 (as old as possible) if they are too recent.
         // This means that they will never be used if either or both are 0.
-        if (crTimestamp > timestamp) crTimestamp = 0;
-        if (pyTimestamp > timestamp) pyTimestamp = 0;
+        if (data.crTimestamp > timestamp) data.crTimestamp = 0;
+        if (data.pyTimestamp > timestamp) data.pyTimestamp = 0;
+
+        // Bounded union prices could have been captured at snapshot that satisfies time constraint.
+        SnapshotSourceLib.Snapshot memory snapshot = SnapshotSourceLib._tryLatestDataAt(
+            boundedUnionSnapshots, boundedAnswer, boundedTimestamp, timestamp, maxTraversal
+        );
+
+        // Update bounded data with constrained source data.
+        (boundedAnswer, boundedTimestamp) = _selectBoundedPrice(
+            data.clAnswer, data.clTimestamp, data.crAnswer, data.crTimestamp, data.pyAnswer, data.pyTimestamp
+        );
+
+        // Return bounded data unless there is a newer snapshotted data that still satisfies time constraint.
+        if (boundedTimestamp >= snapshot.timestamp || snapshot.timestamp > timestamp) {
```

Comment:
If I've followed the logic correctly, should we have `if (boundedTimestamp > snapshot.timestamp)` instead, so we give preference to the snapshot data in case there is one?

I understand that at this point (boundedAnswer, boundedTimestamp) and (snapshot.answer, snapshot.timestamp) could be different values if `boundedTimestamp == snapshot.timestamp`, as the snapshot could have a more complete information?

Reply:
this makes sense, pushed the fix